### PR TITLE
xrdp: init at 0.9.2

### PR DESCRIPTION
**Pull request by volth**
###### Motivation for this change

RDP server (remote desktop which is good for high latency connections).
nix package + nixos module + test script

###### Things done

- [ ] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

### CODE REVIEW

File: nixos/modules/services/networking/xrdp.nix [hunk 1]
```diff
@@ -0,0 +1,150 @@
...
+    
+    substituteInPlace $out/xrdp.ini \
+      --replace "certificate=" "certificate=${cfg.sslCert}" \
+      --replace "key_file=" "key_file=${cfg.sslKey}" \
+      --replace LogFile=xrdp.log LogFile=/dev/null \
+      --replace EnableSyslog=true EnableSyslog=false
+
+    substituteInPlace $out/sesman.ini \
+      --replace LogFile=xrdp-sesman.log LogFile=/dev/null \
+      --replace EnableSyslog=1 EnableSyslog=0
```
Comment (Mic92) [inline nixos/modules/services/networking/xrdp.nix]: Why is syslog disabled? Are logs in journald not useful?

File: pkgs/applications/networking/remote/xrdp/default.nix [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+      substituteInPlace sesman/xauth.c --replace "xauth -q" "${xorg.xauth}/bin/xauth -q"
+      substituteInPlace common/file_loc.h --replace /etc/xrdp $out/etc/xrdp --replace /usr/local $out
+      substituteInPlace instfiles/xrdp.sh --replace /etc/xrdp $out/etc/xrdp --replace /usr/local $out
+    '';
+
+    configurePhase = ''
+      (cd librfxcodec && ./bootstrap && ./configure --prefix=$out --enable-static --disable-shared)
+      ./bootstrap
+      ./configure --prefix=$out --with-systemdsystemunitdir=./do-not-install --enable-ipv6 --enable-jpeg --enable-fuse --enable-rfxcodec --enable-opus
+    '';
```
Comment (Mic92) [inline pkgs/applications/networking/remote/xrdp/default.nix]: Can this phase be splitted into configureFlags and preConfigure?

File: nixos/modules/services/networking/xrdp.nix (hunk 1 above)
Comment (volth) [inline nixos/modules/services/networking/xrdp.nix]: the daemons run in foreground so they duplicate the same log to stdout (and thus to journald)

Comment (Mic92) [inline nixos/modules/services/networking/xrdp.nix]: Ok. In this case it is unnecessary

**Comment by volth** [comment]
When closures were deploying by nixops, ```rsakeys.ini``` was generated once, before the deployment and then was deployed to many machines.
This has a bit of sense of being insecure, so generating ```rsakeys.ini``` postponed until ```systemd``` service start.

File: nixos/modules/services/networking/xrdp.nix [hunk 3]
```diff
@@ -0,0 +1,153 @@
...
+
+    cat > $out/startwm.sh <<EOF
+    #!/bin/sh
+    . /etc/profile
+    ${cfg.defaultWindowManager}
+    EOF
+    chmod +x $out/startwm.sh
+
+    substituteInPlace $out/xrdp.ini \
+      --replace "#rsakeys_ini=" "rsakeys_ini=/var/run/xrdp/rsakeys.ini" \
```
Comment (7c6f434c) [inline nixos/modules/services/networking/xrdp.nix]: Will this result in the keys being regenerated on reboot? Is it desirable (in contrast to storing the keys just in `/var` and allowing trust-on-first-connection logic)?

Comment (volth) [inline nixos/modules/services/networking/xrdp.nix]: Yes, they will be regenerated on reboot.
Unlike the certificates, there is no warning to clients like "the keys have been changed, there is possible a middleman".

**Comment by 7c6f434c** [comment]
The test is not added into the NixOS list of tests to run by default/for channel, but I assume it is intentional

**Comment by volth** [comment]
I am just not familiar with running tests on Hydra.
Perhaps it could make sense to add the test to the list, I do not know.

**Pull request merged**